### ISSUE TITLE
[BUG] stochastic failures of queries involving timestamps with connector athena-jdbc

### ISSUE THREAD

**Issue by kjschiroo** [awslabs/aws-athena-query-federation#362]
**Describe the bug**
Queries involving timestamp columns running against mysql through the jdbc connecter will work the first time, and will generally work the second time, but after repeated runs (usually around 5ish) will begin failing. After the first failure additional failures become more frequent (around 50% failure rate or higher).  

The athena console provides the following error:
```
GENERIC_USER_ERROR: Encountered an exception[java.lang.RuntimeException] from your LambdaFunction[<lambda_arn>] executed in context[S3SpillLocation{bucket=<bucket_name>, key=<key>, directory=true}] with message[java.sql.SQLException: MONTH]
```

**To Reproduce**

1. Setup a mysql database with timestamp columns, some of which are null. (This step is partially conjecture, I haven't created a new database to connect to strictly for debugging this problem.) 
2. Map the database into Athena using the jdbc connector
3. Run a `SELECT * FROM table_name` query. Then run it several more times.

**Expected behavior**
The query should succeed on the nth try in the same way it did on the first try.

**Screenshots / Exceptions / Errors**
Here are the logs out of the lambda function for one of the failures:
```
START RequestId: e33e57d1-3201-4e3c-af06-36f24145e4c5 Version: $LATEST
--
2021-02-28 23:59:21 e33e57d1-3201-4e3c-af06-36f24145e4c5 INFO  JdbcRecordHandler:131 - 96cd0b34-8544-4c1e-9548-40476f06f6e9: Catalog: mysql, table TableName{schemaName=app, tableName=accounts}, splits {partition_name=*}
2021-02-28 23:59:21 e33e57d1-3201-4e3c-af06-36f24145e4c5 INFO  JdbcSplitQueryBuilder:120 - Generated SQL : SELECT `various`, `column`, `names` FROM `app`.`accounts`
2021-02-28 23:59:23  WARN  AmazonS3Client:1903 - No content length specified for stream data.  Stream contents will be buffered in memory and could result in out of memory errors.
2021-02-28 23:59:23  WARN  AmazonS3Client:1903 - No content length specified for stream data.  Stream contents will be buffered in memory and could result in out of memory errors.
2021-02-28 23:59:24  WARN  AmazonS3Client:1903 - No content length specified for stream data.  Stream contents will be buffered in memory and could result in out of memory errors.
2021-02-28 23:59:25 e33e57d1-3201-4e3c-af06-36f24145e4c5 WARN  CompositeHandler:104 - handleRequest: Completed with an exception.java.lang.RuntimeException: java.sql.SQLException: MONTH	at com.amazonaws.athena.connector.lambda.data.S3BlockSpiller.writeRows(S3BlockSpiller.java:177) ~[task/:?]	at com.amazonaws.connectors.athena.jdbc.manager.JdbcRecordHandler.readWithConstraint(JdbcRecordHandler.java:156) ~[task/:?]	at com.amazonaws.connectors.athena.jdbc.MultiplexingJdbcRecordHandler.readWithConstraint(MultiplexingJdbcRecordHandler.java:85) ~[task/:?]	at com.amazonaws.athena.connector.lambda.handlers.RecordHandler.doReadRecords(RecordHandler.java:192) ~[task/:?]	at com.amazonaws.athena.connector.lambda.handlers.RecordHandler.doHandleRequest(RecordHandler.java:158) ~[task/:?]	at com.amazonaws.athena.connector.lambda.handlers.CompositeHandler.handleRequest(CompositeHandler.java:135) ~[task/:?]	at com.amazonaws.athena.connector.lambda.handlers.CompositeHandler.handleRequest(CompositeHandler.java:100) [task/:?]	at lambdainternal.EventHandlerLoader$2.call(EventHandlerLoader.java:902) [LambdaSandboxJava-1.0.jar:?]	at lambdainternal.AWSLambda.startRuntime(AWSLambda.java:340) [LambdaSandboxJava-1.0.jar:?]	at lambdainternal.AWSLambda.<clinit>(AWSLambda.java:63) [LambdaSandboxJava-1.0.jar:?]	at java.lang.Class.forName0(Native Method) ~[?:1.8.0_201]	at java.lang.Class.forName(Class.java:348) [?:1.8.0_201]	at lambdainternal.LambdaRTEntry.main(LambdaRTEntry.java:150) [LambdaJavaRTEntry-1.0.jar:?]Caused by: java.sql.SQLException: MONTH	at com.mysql.cj.jdbc.exceptions.SQLError.createSQLException(SQLError.java:129) ~[task/:?]	at com.mysql.cj.jdbc.exceptions.SQLError.createSQLException(SQLError.java:97) ~[task/:?]	at com.mysql.cj.jdbc.exceptions.SQLError.createSQLException(SQLError.java:89) ~[task/:?]	at com.mysql.cj.jdbc.exceptions.SQLError.createSQLException(SQLError.java:63) ~[task/:?]	at com.mysql.cj.jdbc.exceptions.SQLError.createSQLException(SQLError.java:73) ~[task/:?]	at com.mysql.cj.jdbc.exceptions.SQLExceptionsMapping.translateException(SQLExceptionsMapping.java:85) ~[task/:?]	at com.mysql.cj.jdbc.result.ResultSetImpl.getTimestamp(ResultSetImpl.java:903) ~[task/:?]	at com.mysql.cj.jdbc.result.ResultSetImpl.getTimestamp(ResultSetImpl.java:948) ~[task/:?]	at com.amazonaws.connectors.athena.jdbc.manager.JdbcRecordHandler.lambda$makeExtractor$13(JdbcRecordHandler.java:266) ~[task/:?]	at com.amazonaws.athena.connector.lambda.data.writers.fieldwriters.DateMilliFieldWriter.write(DateMilliFieldWriter.java:78) ~[task/:?]	at com.amazonaws.athena.connector.lambda.data.writers.GeneratedRowWriter.writeRow(GeneratedRowWriter.java:116) ~[task/:?]	at com.amazonaws.connectors.athena.jdbc.manager.JdbcRecordHandler.lambda$readWithConstraint$0(JdbcRecordHandler.java:156) ~[task/:?]	at com.amazonaws.athena.connector.lambda.data.S3BlockSpiller.writeRows(S3BlockSpiller.java:174) ~[task/:?]	... 12 moreCaused by: com.mysql.cj.exceptions.WrongArgumentException: MONTH	at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method) ~[?:1.8.0_201]	at sun.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62) ~[?:1.8.0_201]	at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45) ~[?:1.8.0_201]	at java.lang.reflect.Constructor.newInstance(Constructor.java:423) ~[?:1.8.0_201]	at com.mysql.cj.exceptions.ExceptionFactory.createException(ExceptionFactory.java:61) ~[task/:?]	at com.mysql.cj.exceptions.ExceptionFactory.createException(ExceptionFactory.java:105) ~[task/:?]	at com.mysql.cj.result.SqlTimestampValueFactory.localCreateFromTimestamp(SqlTimestampValueFactory.java:112) ~[task/:?]	at com.mysql.cj.result.SqlTimestampValueFactory.localCreateFromTimestamp(SqlTimestampValueFactory.java:50) ~[task/:?]	at com.mysql.cj.result.AbstractDateTimeValueFactory.createFromTimestamp(AbstractDateTimeValueFactory.java:87) ~[task/:?]	at com.mysql.cj.protocol.a.MysqlTextValueDecoder.decodeTimestamp(MysqlTextValueDecoder.java:79) ~[task/:?]	at com.mysql.cj.protocol.result.AbstractResultsetRow.decodeAndCreateReturnValue(AbstractResultsetRow.java:87) ~[task/:?]	at com.mysql.cj.protocol.result.AbstractResultsetRow.getValueFromBytes(AbstractResultsetRow.java:241) ~[task/:?]	at com.mysql.cj.protocol.a.result.TextBufferRow.getValue(TextBufferRow.java:132) ~[task/:?]	at com.mysql.cj.jdbc.result.ResultSetImpl.getTimestamp(ResultSetImpl.java:903) ~[task/:?]	at com.mysql.cj.jdbc.result.ResultSetImpl.getTimestamp(ResultSetImpl.java:948) ~[task/:?]	at com.amazonaws.connectors.athena.jdbc.manager.JdbcRecordHandler.lambda$makeExtractor$13(JdbcRecordHandler.java:266) ~[task/:?]	at com.amazonaws.athena.connector.lambda.data.writers.fieldwriters.DateMilliFieldWriter.write(DateMilliFieldWriter.java:78) ~[task/:?]	at com.amazonaws.athena.connector.lambda.data.writers.GeneratedRowWriter.writeRow(GeneratedRowWriter.java:116) ~[task/:?]	at com.amazonaws.connectors.athena.jdbc.manager.JdbcRecordHandler.lambda$readWithConstraint$0(JdbcRecordHandler.java:156) ~[task/:?]	at com.amazonaws.athena.connector.lambda.data.S3BlockSpiller.writeRows(S3BlockSpiller.java:174) ~[task/:?]	... 12 moreCaused by: java.lang.IllegalArgumentException: MONTH	at java.util.GregorianCalendar.computeTime(GregorianCalendar.java:2648) ~[?:1.8.0_201]	at java.util.Calendar.updateTime(Calendar.java:3393) ~[?:1.8.0_201]	at java.util.Calendar.getTimeInMillis(Calendar.java:1782) ~[?:1.8.0_201]	at com.mysql.cj.result.SqlTimestampValueFactory.localCreateFromTimestamp(SqlTimestampValueFactory.java:108) ~[task/:?]	at com.mysql.cj.result.SqlTimestampValueFactory.localCreateFromTimestamp(SqlTimestampValueFactory.java:50) ~[task/:?]	at com.mysql.cj.result.AbstractDateTimeValueFactory.createFromTimestamp(AbstractDateTimeValueFactory.java:87) ~[task/:?]	at com.mysql.cj.protocol.a.MysqlTextValueDecoder.decodeTimestamp(MysqlTextValueDecoder.java:79) ~[task/:?]	at com.mysql.cj.protocol.result.AbstractResultsetRow.decodeAndCreateReturnValue(AbstractResultsetRow.java:87) ~[task/:?]	at com.mysql.cj.protocol.result.AbstractResultsetRow.getValueFromBytes(AbstractResultsetRow.java:241) ~[task/:?]	at com.mysql.cj.protocol.a.result.TextBufferRow.getValue(TextBufferRow.java:132) ~[task/:?]	at com.mysql.cj.jdbc.result.ResultSetImpl.getTimestamp(ResultSetImpl.java:903) ~[task/:?]	at com.mysql.cj.jdbc.result.ResultSetImpl.getTimestamp(ResultSetImpl.java:948) ~[task/:?]	at com.amazonaws.connectors.athena.jdbc.manager.JdbcRecordHandler.lambda$makeExtractor$13(JdbcRecordHandler.java:266) ~[task/:?]	at com.amazonaws.athena.connector.lambda.data.writers.fieldwriters.DateMilliFieldWriter.write(DateMilliFieldWriter.java:78) ~[task/:?]	at com.amazonaws.athena.connector.lambda.data.writers.GeneratedRowWriter.writeRow(GeneratedRowWriter.java:116) ~[task/:?]	at com.amazonaws.connectors.athena.jdbc.manager.JdbcRecordHandler.lambda$readWithConstraint$0(JdbcRecordHandler.java:156) ~[task/:?]	at com.amazonaws.athena.connector.lambda.data.S3BlockSpiller.writeRows(S3BlockSpiller.java:174) ~[task/:?]	... 12 more
java.sql.SQLException: MONTH: java.lang.RuntimeExceptionjava.lang.RuntimeException: java.sql.SQLException: MONTH	at com.amazonaws.athena.connector.lambda.data.S3BlockSpiller.writeRows(S3BlockSpiller.java:177)	at com.amazonaws.connectors.athena.jdbc.manager.JdbcRecordHandler.readWithConstraint(JdbcRecordHandler.java:156)	at com.amazonaws.connectors.athena.jdbc.MultiplexingJdbcRecordHandler.readWithConstraint(MultiplexingJdbcRecordHandler.java:85)	at com.amazonaws.athena.connector.lambda.handlers.RecordHandler.doReadRecords(RecordHandler.java:192)	at com.amazonaws.athena.connector.lambda.handlers.RecordHandler.doHandleRequest(RecordHandler.java:158)	at com.amazonaws.athena.connector.lambda.handlers.CompositeHandler.handleRequest(CompositeHandler.java:135)	at com.amazonaws.athena.connector.lambda.handlers.CompositeHandler.handleRequest(CompositeHandler.java:100)Caused by: java.sql.SQLException: MONTH	at com.mysql.cj.jdbc.exceptions.SQLError.createSQLException(SQLError.java:129)	at com.mysql.cj.jdbc.exceptions.SQLError.createSQLException(SQLError.java:97)	at com.mysql.cj.jdbc.exceptions.SQLError.createSQLException(SQLError.java:89)	at com.mysql.cj.jdbc.exceptions.SQLError.createSQLException(SQLError.java:63)	at com.mysql.cj.jdbc.exceptions.SQLError.createSQLException(SQLError.java:73)	at com.mysql.cj.jdbc.exceptions.SQLExceptionsMapping.translateException(SQLExceptionsMapping.java:85)	at com.mysql.cj.jdbc.result.ResultSetImpl.getTimestamp(ResultSetImpl.java:903)	at com.mysql.cj.jdbc.result.ResultSetImpl.getTimestamp(ResultSetImpl.java:948)	at com.amazonaws.connectors.athena.jdbc.manager.JdbcRecordHandler.lambda$makeExtractor$13(JdbcRecordHandler.java:266)	at com.amazonaws.athena.connector.lambda.data.writers.fieldwriters.DateMilliFieldWriter.write(DateMilliFieldWriter.java:78)	at com.amazonaws.athena.connector.lambda.data.writers.GeneratedRowWriter.writeRow(GeneratedRowWriter.java:116)	at com.amazonaws.connectors.athena.jdbc.manager.JdbcRecordHandler.lambda$readWithConstraint$0(JdbcRecordHandler.java:156)	at com.amazonaws.athena.connector.lambda.data.S3BlockSpiller.writeRows(S3BlockSpiller.java:174)	... 6 moreCaused by: com.mysql.cj.exceptions.WrongArgumentException: MONTH	at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)	at sun.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)	at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)	at java.lang.reflect.Constructor.newInstance(Constructor.java:423)	at com.mysql.cj.exceptions.ExceptionFactory.createException(ExceptionFactory.java:61)	at com.mysql.cj.exceptions.ExceptionFactory.createException(ExceptionFactory.java:105)	at com.mysql.cj.result.SqlTimestampValueFactory.localCreateFromTimestamp(SqlTimestampValueFactory.java:112)	at com.mysql.cj.result.SqlTimestampValueFactory.localCreateFromTimestamp(SqlTimestampValueFactory.java:50)	at com.mysql.cj.result.AbstractDateTimeValueFactory.createFromTimestamp(AbstractDateTimeValueFactory.java:87)	at com.mysql.cj.protocol.a.MysqlTextValueDecoder.decodeTimestamp(MysqlTextValueDecoder.java:79)	at com.mysql.cj.protocol.result.AbstractResultsetRow.decodeAndCreateReturnValue(AbstractResultsetRow.java:87)	at com.mysql.cj.protocol.result.AbstractResultsetRow.getValueFromBytes(AbstractResultsetRow.java:241)	at com.mysql.cj.protocol.a.result.TextBufferRow.getValue(TextBufferRow.java:132)	... 13 moreCaused by: java.lang.IllegalArgumentException: MONTH	at java.util.GregorianCalendar.computeTime(GregorianCalendar.java:2648)	at java.util.Calendar.updateTime(Calendar.java:3393)	at java.util.Calendar.getTimeInMillis(Calendar.java:1782)	at com.mysql.cj.result.SqlTimestampValueFactory.localCreateFromTimestamp(SqlTimestampValueFactory.java:108)	... 19 more
END RequestId: e33e57d1-3201-4e3c-af06-36f24145e4c5
REPORT RequestId: e33e57d1-3201-4e3c-af06-36f24145e4c5	Duration: 4250.30 ms	Billed Duration: 4251 ms	Memory Size: 3008 MB	Max Memory Used: 489 MB

```

**Connector Details (please complete the following information):**
 - Version: 2021.6.1
 - Name: jdbc (named `mysql` in Athena).
 - Athena Query IDs: `34f1eec1-5ef7-4e21-8243-e01ff0701877` (One of the failed queries)

**Comment by kjschiroo** [comment]
Is there an environment variable I can set to get more verbose logging?

**Comment by kjschiroo** [comment]
Dear persons of the future,

This problem was resolved so here are a couple notes that may help you debug it in your case:

- The stochastic nature of it was caused by the query not having a specific ordering specified. Once an `ORDER BY` was specified the problem became consistent. 
- One of the key things to understand is that the connector / Athena will usually query significantly more rows that it should logically need to, so the row that was causing issues may be significantly removed from the rows that you're actually trying to get back from Athena. In my case I'd had a `LIMIT 1` on my query that led me to believe the issue was with the very first row in the database which seemed fine, but the actual problem was 31563 rows later. I ended up performing a binary search on the problem table to figure out where the exact row was. At least in my situation having a `WHERE id < <number>` was pushed down into the connector allowing me to figure out where the problem row was.
- `mysql` is far more flexible about timestamps than this connector is willing to be, in this case mysql was willing to tolerate a pretty incoherent timestamp (2018-00-15 04:59:59).

Hope that helps!
